File: src/ingest/claude-code.ts
import {
  contentToText,
  findJsonlFiles,
  type FileContext,
  type ParsedMessage,
  type TranscriptParser,
} from "./parser.js";

// Parses Claude Code transcripts under ~/.claude/projects/<slug>/*.jsonl
// (including subagents/). Each record is self-describing — it carries its
// own uuid, sessionId, cwd, and timestamp — so no cross-line state is
// needed, but ctx is still refreshed for consistency with formats that do.
//
// A record's message.content is either a plain string (a user turn) or an
// array of blocks (text / thinking / tool_use for assistant turns,
// tool_result for tool returns). One record therefore expands to several
// messages; block index disambiguates their record ids.
//
// The format is unversioned and carries many non-conversation record types
// (queue-operation, attachment, last-prompt, file-history-snapshot,
// ai-title, …); anything that is not user/assistant/system is skipped.

interface ClaudeRecord {
  type?: string;
  uuid?: string;
  sessionId?: string;
  cwd?: string;
  timestamp?: string;
  message?: { role?: string; content?: unknown };
}

const CONVERSATION_TYPES = new Set(["user", "assistant", "system"]);

export const claudeCodeParser: TranscriptParser = {
  format: "claude-code",

  enumerate(dirs: string[]): string[] {
    return findJsonlFiles(dirs);
  },

  parse(line: string, ctx: FileContext): ParsedMessage[] {
    const trimmed = line.trim();
    if (trimmed === "") return [];
    let record: ClaudeRecord;
    try {
      record = JSON.parse(trimmed) as ClaudeRecord;
    } catch {
      return [];
    }
    if (!record.type || !CONVERSATION_TYPES.has(record.type)) return [];

    if (typeof record.sessionId === "string") ctx.sessionId = record.sessionId;
    if (typeof record.cwd === "string") ctx.projectPath = record.cwd;
    const sessionId = ctx.sessionId;
    const recordId = record.uuid;
    if (!sessionId || !recordId) return []; // cannot key a message without both

    const common = {
      nativeSessionId: sessionId,
      ...(record.timestamp ? { nativeTs: record.timestamp } : {}),
      ...(ctx.projectPath ? { projectPath: ctx.projectPath } : {}),
    };

    if (record.type === "system") {
      const text = contentToText(record.message?.content ?? (record as { content?: unknown }).content);
      if (text === "") return [];
      return [{ ...common, nativeRecordId: recordId, role: "system", kind: "system", content: text }];
    }

    const role = record.message?.role ?? record.type;
    const content = record.message?.content;

    if (typeof content === "string") {
      if (content === "") return [];
      return [{ ...common, nativeRecordId: recordId, role, kind: "message", content }];
    }
    if (!Array.isArray(content)) return [];

    const out: ParsedMessage[] = [];
    content.forEach((block, i) => {
      const parsed = parseBlock(block, role);
      if (parsed) out.push({ ...common, nativeRecordId: `${recordId}#${i}`, ...parsed });
    });
    return out;
  },
};

/** Maps one content block to {role, kind, content}, or null to skip. */
function parseBlock(
  block: unknown,
  recordRole: string,
): Pick<ParsedMessage, "role" | "kind" | "content"> | null {
  if (!block || typeof block !== "object") return null;
  const b = block as Record<string, unknown>;
  switch (b["type"]) {
    case "text": {
      const text = typeof b["text"] === "string" ? b["text"] : "";
      return text === "" ? null : { role: recordRole, kind: "message", content: text };
    }
    case "thinking": {
      const text = typeof b["thinking"] === "string" ? b["thinking"] : "";
      return text === "" ? null : { role: recordRole, kind: "reasoning", content: text };
    }
    case "tool_use":
      return {
        role: recordRole,
        kind: "tool_use",
        content: JSON.stringify({ id: b["id"], name: b["name"], input: b["input"] }),
      };
    case "tool_result":
      return {
        role: "tool",
        kind: "tool_result",
        content: JSON.stringify({
          tool_use_id: b["tool_use_id"],
          is_error: b["is_error"] ?? false,
          content: contentToText(b["content"]),
        }),
      };
    default:
      return null;
  }
}

File: src/ingest/claude_code.rs
//! Parses Claude Code transcripts under ~/.claude/projects/<slug>/*.jsonl
//! (including subagents/). Each record is self-describing — it carries its
//! own uuid, sessionId, cwd, and timestamp — so no cross-line state is
//! needed, but ctx is still refreshed for consistency with formats that do.
//!
//! A record's message.content is either a plain string (a user turn) or an
//! array of blocks (text / thinking / tool_use for assistant turns,
//! tool_result for tool returns). One record therefore expands to several
//! messages; block index disambiguates their record ids.
//!
//! The format is unversioned and carries many non-conversation record types
//! (queue-operation, attachment, last-prompt, file-history-snapshot,
//! ai-title, …); anything that is not user/assistant/system is skipped.

use std::path::PathBuf;

use serde_json::{Value, json};

use super::parser::{
    FileContext, ParsedMessage, TranscriptParser, content_to_text, find_jsonl_files, json_of_keys,
};
use crate::js;

const CONVERSATION_TYPES: [&str; 3] = ["user", "assistant", "system"];

pub struct ClaudeCodeParser;

impl TranscriptParser for ClaudeCodeParser {
    fn format(&self) -> &'static str {
        "claude-code"
    }

    fn enumerate(&self, dirs: &[PathBuf]) -> Vec<PathBuf> {
        find_jsonl_files(dirs)
    }

    fn parse(&self, line: &str, ctx: &mut FileContext) -> Vec<ParsedMessage> {
        let trimmed = js::trim(line);
        let Ok(record) = serde_json::from_str::<Value>(trimmed) else { return vec![] };
        let Some(record_type) = record.get("type").and_then(Value::as_str) else { return vec![] };
        if !CONVERSATION_TYPES.contains(&record_type) {
            return vec![];
        }

        if let Some(session_id) = record.get("sessionId").and_then(Value::as_str) {
            ctx.session_id = Some(session_id.to_string());
        }
        if let Some(cwd) = record.get("cwd").and_then(Value::as_str) {
            ctx.project_path = Some(cwd.to_string());
        }
        let (Some(session_id), Some(record_id)) =
            (ctx.session_id.clone(), record.get("uuid").and_then(Value::as_str))
        else {
            return vec![]; // cannot key a message without both
        };
        let message = |record_id: String, role: &str, kind: &str, content: String| ParsedMessage {
            native_session_id: session_id.clone(),
            native_record_id: record_id,
            role: role.to_string(),
            kind: kind.to_string(),
            content,
            native_ts: record
                .get("timestamp")
                .and_then(Value::as_str)
                .filter(|t| !t.is_empty())
                .map(str::to_string),
            project_path: ctx.project_path.clone().filter(|p| !p.is_empty()),
        };

        let message_field = record.get("message");
        if record_type == "system" {
            // `message.content ?? content`: older system records carry it at the top level.
            let content = message_field
                .and_then(|m| m.get("content"))
                .filter(|c| !c.is_null())
                .or_else(|| record.get("content"));
            let text = content.map(content_to_text).unwrap_or_default();
            if text.is_empty() {
                return vec![];
            }
            return vec![message(record_id.to_string(), "system", "system", text)];
        }

        let role = message_field
            .and_then(|m| m.get("role"))
            .and_then(Value::as_str)
            .unwrap_or(record_type);
        match message_field.and_then(|m| m.get("content")) {
            Some(Value::String(content)) => {
                if content.is_empty() {
                    vec![]
                } else {
                    vec![message(record_id.to_string(), role, "message", content.clone())]
                }
            }
            Some(Value::Array(blocks)) => blocks
                .iter()
                .enumerate()
                .filter_map(|(i, block)| {
                    let (block_role, kind, content) = parse_block(block, role)?;
                    Some(message(format!("{record_id}#{i}"), &block_role, kind, content))
                })
                .collect(),
            _ => vec![],
        }
    }
}

/// Maps one content block to (role, kind, content), or None to skip.
fn parse_block(block: &Value, record_role: &str) -> Option<(String, &'static str, String)> {
    let text_or_skip = |key: &str, kind: &'static str| {
        let text = block.get(key).and_then(Value::as_str).unwrap_or_default();
        if text.is_empty() { None } else { Some((record_role.to_string(), kind, text.to_string())) }
    };
    match block.get("type").and_then(Value::as_str)? {
        "text" => text_or_skip("text", "message"),
        "thinking" => text_or_skip("thinking", "reasoning"),
        "tool_use" => Some((
            record_role.to_string(),
            "tool_use",
            json_of_keys(block, &["id", "name", "input"]),
        )),
        "tool_result" => {
            let mut result = serde_json::Map::new();
            if let Some(id) = block.get("tool_use_id") {
                result.insert("tool_use_id".into(), id.clone());
            }
            let is_error =
                block.get("is_error").filter(|v| !v.is_null()).cloned().unwrap_or(json!(false));
            result.insert("is_error".into(), is_error);
            let content = block.get("content").map(content_to_text).unwrap_or_default();
            result.insert("content".into(), Value::String(content));
            Some(("tool".to_string(), "tool_result", Value::Object(result).to_string()))
        }
        _ => None,
    }
}
